Add attestation spec (#4)
openvex/spec: ATTESTING.md

# Attesting OpenVEX Documents

## What is an Attestation?

An attestation is an assertion made about a piece of software. There are many

> kinds of attestations in use today such as provenance attestation defined by
> SLSA or those attesting to the results of vulnerability scans. 

Reviewer: ```suggestion
kinds of attestations in use today, such as the "provenance" attestation defined by
SLSA, or attestations of vulnerability scan results. 
```

Author: I was trying to avoid repeating "attestation(s)" so much, changed to:

> or those asserting to the results of vulnerability scans

OpenVEX was conceived to be able to be embedded in 
[in-toto attestations](https://github.com/in-toto/attestation). The format defined
by the in-toto project is composed of a number of subjects (the pieces of 
software the attestation is talking about) and a predicate that defines what is
being said about the subjects.

[DIAGRAM]

> Here is an example of an empty attestation

Reviewer: nit: I think "attestation" typically refers to the envelope, and what's shown here is referred to as the "statement" (or "in-toto statement" sometimes). But let's not block on this right now, maybe something to ponder afterward.

Author: OK, lets discuss this one further

```json
{
  "_type": "https://in-toto.io/Statement/v0.1",
  "predicateType": "",
  "subject": [], 
  "predicate": {}
}
```

## Embedding and Inheritance

OpenVEX documents are designed to be embeddable in other formats. This is not a

> unique feature of OpenVEX, the VEX minimum elements define the notion of an

Reviewer: ```suggestion
unique feature of OpenVEX: the VEX minimum elements define the notion of an
```

"encapsulating format", a document that contains the VEX document and its
statements. VEX also defines an inheritance model where the required data to 
complete VEX metadata cascades down from the encapsulating format to the 

> document, to the statement. This allows VEX to leverage the capabilities of the

Reviewer: ```suggestion
document, and then to the statement. This allows VEX to leverage the capabilities of the
```

encapsulating formats while defining a compatibility flow among implementations.

OpenVEX documents do not require an encapsulating document. Nevertheless, they
were designed to be embeddable and they can be used as in-toto predicates. This
lets software authors assert VEX data about a piece of software. 

When embedding OpenVEX in attestations, the only field of data that "cascades"

> is the VEX statement's `product` or `subject` in in-toto lingo. 

Reviewer: ```suggestion
is the VEX statement's `product`, or `subject` in in-toto lingo. 
```

      "products": [
        "pkg:apk/distro/git@2.39.0-r1?arch=armv7",
        "pkg:apk/distro/git@2.39.0-r1?arch=x86_64"
      ],
      "status": "fixed"
    }
```

In the previous example, the statement specifies two packages of git (for armv7 and x86_64).

Attestations define their subjects natively. Following the VEX inheritance model
the subjects in an attestation containing a VEX document will cascade down and
become the VEX "product" of any statements that don't specify a subject.

Here is an example of an attestation with the same packages as subjects. Note that
the predicate type is set to the OpenVEX context (the predicate contents have been
removed for clarity):

```json
{
  "_type": "https://in-toto.io/Statement/v0.1",
  "predicateType": "https://openvex.dev/ns",
  "subject": [
    {
      "name": "pkg:apk/distro/git@2.39.0-r1?arch=armv7",
      "digest": {
        "sha256": "74634d9736a45ca9f6e1187e783492199e020f4a5c19d0b1abc2b604f894ac99"
      },
    },
    {
      "name": "pkg:apk/distro/git@2.39.0-r1?arch=x86_64",
      "digest": {
        "sha256": "6bd98fe56e4d91439343d123d98522005874957ea1cb6075e75544d7753bd8d7"
      },
    }
  ],
  "predicate": {

>     ...

Reviewer: ```suggestion
    // ...
```

(super nit — just if we want to avoid the bright red "this isn't valid JSON" highlights)

Author: yeah I hate that. fixed

Reviewer: Same :D I think the code block type might need to be `jsonc`, too, to allow comments 🤔

  }
}

```

When embedding OpenVEX inside an attestation, the subjects SHOULD move from the

> vex statement product to the attestation subjects. This makes the attestation usable

Reviewer: ```suggestion
VEX statement product to the attestation subjects. This makes the attestation usable
```

for consistency

subjects' digests can be computed externally.

The product entries MAY remain in the VEX statement. In that case, they MUST
be repeated and matched in the attestation subject section. An attestation SHOULD
remain complete when composed with an OpenVEX predicate.

## Handling Product/Subject Granularity

An attestation's predicate is a singleton. It is a set of exactly one predicate
that applies to any number of subjects. VEX, on the other hand, defines a document
model that can host any number of statements, possibly with different subjects:

```json
"statements": [
    {
      "vulnerability": "CVE-2014-123456",
      "products": [
        "pkg:apk/distro/git@2.39.0-r1?arch=armv7"
      ],
      "status": "fixed"
    },
    {
      "vulnerability": "CVE-2014-123456",
      "products": [
        "pkg:apk/distro/git@2.39.0-r1?arch=x86_64"
      ],
      "status": "under_investigation"
    },
]
```

The nature of the data models implies that an attestation can only refer to 
VEX statements that contain one or more of the `subject` entries in their product
section. To attest the example above, an attestation can use any of the following
subject structs:

```json

>     subject: {

Reviewer: ```suggestion
    "subject": {
```

JSON object fields have to be quoted

Also — and please **ignore this** if these examples aren't meant to be taken literally — in-toto subjects are array types, not object types. And each array element is an object with a `name` and `digest` field, both of which are required.

Ditto this comment for the following subject examples, too

        "pkg:apk/distro/git@2.39.0-r1?arch=armv7"
    }
```

```json
    subject: {
        "pkg:apk/distro/git@2.39.0-r1?arch=x86_64"
    }
```

```json
    subject: {
        "pkg:apk/distro/git@2.39.0-r1?arch=armv7",
        "pkg:apk/distro/git@2.39.0-r1?arch=x86_64"
    }
```

When an identifier is listed in the subject section, it signals any processor
to look for data about it in the predicate. When looking at the VEX statements,
the following rules define how statements are to be considered:

1. Any VEX statements that don't define products are considered to be attested.
2. Any VEX statements listing products but not having the attestation's subjects

> in the product list are to be ignored

Reviewer: ```suggestion
in the product list are to be ignored.
```

3. Any VEX statements including one of the attestation's subjects in its `product`
section are to be considered only for that identifier and others that match.
4. If a VEX statement lists one of the attestation's subjects in the product list
but not another, it MUST be considered for the former but not for the latter.

## Digital Signatures

Attestations are meant to be digitally signed. While the signature envelope is

> not a part of the attestation or OpenVEX specifications, it should be noted that

Reviewer: nit: I think the envelope is actually [defined in in-toto](https://github.com/in-toto/attestation/blob/main/spec/README.md#envelope)

Author: IIRC they gave up having their own envelope instead choosing to recommend DSSE. We can punt this one for later.
